content/docs/rfcs/24/README.md

```diff
@@ -0,0 +1,60 @@
+---
+slug: 24
+title: 24/WAKU-VOTING
+name: Waku v2 Voting
+status: raw
+editor: Szymon Szlachtowicz <szymon.s@ethworks.io>
+---
+
+This spec is a proposition for a voting protocol over Waku V2.
+
+# Motivation
+
+In open p2p protocol there is an issue with voting off-chain as there is much room for malicious peers to only include votes that support their case when submitting votes to chain.
+
+Proposed solution is to agregate votes over waku and allow users to submit votes to smart contract that aren't already submitted.
+
+# Smart contract
+
+Voting should be finalized on chain so that the finished vote is immutable. Becouse of that, smart contract needs to be deployed. New smart contract will have to be able to start a vote and accept submited votes.
+
+## Double voting
+
+Smart contract should also keep a list of all votes so that no one can double vote.
+
+# Voting
+
+## Sending votes
+
+Sending votes is simple every peer is able to send a message to Waku topic specific to given application: 
+```
+/{application-name}/{voting-room}
+```
+
+vote object that is sent over waku should contain information about: 
+
+```ts
+{
+    sender: String //address of a sender
+    vote: String //vote sent eg. 'yes' 'no'
+    sntAmount: BigNumber //number of snt cast on vote
+    sign: String // cryptographic sign of a transaction
+    nonce: Number // number of votes cast from this address on current vote
+    room: String // address of a voting room
+}
+```
+
+## Aggregating votes
+
+Every peer that is opening specific voting room will listen to votes sent over p2p network, and aggregate them for a single transaction to chain
```

Comment:
```suggestion
Every peer that is opening specific voting room will listen to votes sent over p2p network, and aggregate them for a single transaction to chain.
```

What we discussed was:
- Alice open vote for room 1, she will vote yes so she listens to aggregate yes votes and publshes her yes vote on Waku.
- Bob is made aware of vote for room 1, he votes no, he publishes his no vote on Waku, retrieve previous no votes and starts listening to no votes.
- Carole is made aware of vote for room 1, she votes yes, she publishes her yes vote on Waku, retrieve previous yes votes and starts listening to yes votes.
- Bob really does not want this community to be in the directory, he aggregates no votes and submit to the blockchain
- Carole is really excited about this community, she aggregates yes votes and submit to the blockchain

Reply:
Yes but as someone pointed out i think it was Marek typical user isn't really aware of what is going under the hood and he doesn't need to know if he is aggregating all the votes or only no or yes votes, he only needs to know if aggregated votes would change the outcome.

 In your example above Bob would see that submitting votes to blockchain would be against his wishes so he doesn't do that but Carole sees that if she submits the votes she can change the results in her favour so she submits votes for and against because she is a good user. That way we can save some gas for users, and create a simpler aggregator.

Now if someone wanted to be malicious and only aggregate votes against normal user would see that only votes for are left and he can simply submit all the votes he has aggregated which are votes for.

Reply:
> if aggregated votes would change the outcome.

So yes, we discussed that.

I think if we are adding conditionals such as "only show to the user the option to submit if it can change the outcome" we could end up degrading the UX.

Maybe a simplest way to proceed would be to simply show to the user:
"4000 SNT voted yes, 3000 voted no. There are 1000 pending yes vote. Do you want to submit them?".

This way, we don't have to fiddle with conditionals based on number of votes and last minute voting, etc.

Reply:
So maybe we could just show current submitted votes and under this show how the votes would look like after submission for example:

"4000 SNT voted yes, 3000 SNT voted no. There are pending votes. Do you want to submit them?".

And if someone would click submit button modal would show up with information how votes would look like after submission:

"Votes after submission 5000 SNT yes, 6000 SNT no."

This way we have simple logic of aggregating all votes with no need to track what user voted before and filtering votes